dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -216,86 +220,110 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationActivators const& activators)
     {
         try
         {
-            THROW_HR_IF_NULL(E_INVALIDARG, token);
-            if (WI_IsFlagSet(options, PushNotificationRegistrationOptions::PushTrigger))
+            auto lock = s_activatorInfoLock.lock_exclusive();
+            if (WI_IsFlagSet(activators, PushNotificationRegistrationActivators::PushTrigger))
             {
-                auto taskRegistration = token.TaskRegistration();
-                THROW_HR_IF_NULL(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), taskRegistration);
-                taskRegistration.Unregister(true);
+                THROW_HR_IF_NULL_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), s_pushTriggerRegistration, "PushTrigger not registered.");
+                s_pushTriggerRegistration.Unregister(true);
+                s_pushTriggerRegistration = nullptr;
             }
 
-            // Check for COM flag, a valid cookie and if there are outstanding locks on the PushNotificationBackgroundTask class factory
-            if (WI_IsFlagSet(options, PushNotificationRegistrationOptions::ComActivator) && token.Cookie() && winrt::get_module_lock() == 0)
+            // Check for COM flag, a valid cookie
+            if (WI_IsFlagSet(activators, PushNotificationRegistrationActivators::ComActivator))
             {
-                LOG_IF_FAILED(::CoRevokeClassObject(static_cast<DWORD>(token.Cookie())));
+                THROW_HR_IF_MSG(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), !s_comActivatorRegistration, "ComActivator not registered.");
+                s_comActivatorRegistration.reset();
             }
         }
-
         catch (...)
         {
             HRESULT hrError = wil::ResultFromCaughtException();
-            PushNotificationTelemetry::ActivatorUnregisteredByApi(hrError, options);
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(hrError, activators);
             THROW_HR(hrError);
         }
 
-        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK, options);
+        PushNotificationTelemetry::ActivatorUnregisteredByApi(S_OK, activators);
+    }
+
+    void PushNotificationManager::UnregisterAllActivators()
+    {
+        try
+        {
+            auto lock = s_activatorInfoLock.lock_exclusive();
+            if (s_pushTriggerRegistration)
+            {
+                s_pushTriggerRegistration.Unregister(true);
+                s_pushTriggerRegistration = nullptr;
+            }
+
+            s_comActivatorRegistration.reset();
+        }
+        catch(...)
+        {
+            HRESULT hrError = wil::ResultFromCaughtException();
+            PushNotificationTelemetry::ActivatorUnregisteredByApi(hrError, PushNotificationRegistrationActivators::PushTrigger | PushNotificationRegistrationActivators::ComActivator);
```

Comment:
Does the OSS version of WIL's telemetry helpers support creating Activities?  Those can watch the thread for thrown exceptions and log on them, rather than you having to wrap everything in try/catch(e){log(e);throw;} sequences.

Reply:
Captured here: https://github.com/microsoft/WindowsAppSDK/issues/1391